server/use_cases/add_service_commitments.py

```diff
@@ -39,52 +41,99 @@ def add_service_commitments(commitments_repo, shifts_repo, commitments):
         else:
             valid_commitments.append(commitment)
             valid_indexes.append(i)
-            results.append(None)  # placeholder for now
-
+            results.append(None)
     if not valid_commitments:
         return results
-
-    # check for time overlap in valid shifts
+    # Get user ID from first commitment
+    user_id = valid_commitments[0].volunteer_id
+    # Fetch user's existing commitments
+    try:
+        existing_user_commitments = commitments_repo.fetch_service_commitments(
+            user_id=user_id)
+        if existing_user_commitments is None:
+            existing_user_commitments = []
+    except (AttributeError, TypeError):
+        existing_user_commitments = []
+    existing_commitment_shift_ids = [str(c.service_shift_id)
+                                    for c in existing_user_commitments
+                                    if c.service_shift_id is not None]
+    # Get existing shifts
+    existing_user_shifts = []
+    if existing_commitment_shift_ids:
+        existing_user_shifts = shifts_repo.get_shifts(
+            existing_commitment_shift_ids)
+    # Get valid shifts for current commitments
     valid_shifts = [
-        existing_shifts[c.service_shift_id]
-        for c in valid_commitments
-    ]
-    if check_time_overlap(valid_shifts):
-        for i in valid_indexes:
-            results[i] = {
+        existing_shifts[c.service_shift_id] for c in valid_commitments]
+    final_valid_commitments = []
+    final_valid_indexes = []
+    for i, idx in enumerate(valid_indexes):
+        current_shift = valid_shifts[i]
+        has_overlap = False
+        # Check for overlap with existing user shifts
+        shifts_to_check = [current_shift]
+        shifts_to_check.extend(existing_user_shifts)
+        if check_time_overlap(shifts_to_check):
+            has_overlap = True
+        if not has_overlap and final_valid_indexes:
+            approved_shifts = [
+                valid_shifts[j] for j in range(i) if j in final_valid_indexes]
+            shifts_to_check = [current_shift]
+            shifts_to_check.extend(approved_shifts)
+            if check_time_overlap(shifts_to_check):
+                has_overlap = True
```

Comment:
```suggestion
```
It's redundant to do the check for overlap again. I understand why you are doing it - you want to make sure that it doesn't overlap with any of the commitments sent in the same request that you already approved. There is a simpler way to do this. See my suggestion later in this function.